backend/tests/e2e/test_socketio.py

```diff
@@ -0,0 +1,138 @@
+from typing import List, Dict
+
+import pytest
+import socketio
+
+SERVER_URL = "http://localhost:8082"
+
+
+class MockClient:
+    def __init__(self, nickname: str, toplvl_namespace="/game"):
+        self._sio_client = socketio.AsyncSimpleClient(logger=False)
+        self._toplvl_namespace = toplvl_namespace
+        self._nickname = nickname
+        self._msg_log: List[Dict] = []
+
+    @property
+    def sio(self):
+        return self._sio_client
+
+    @property
+    def nickname(self):
+        return self._nickname
+
+    @property
+    def messages_log(self):
+        return self._msg_log
+
+    async def connect(self, url: str):
+        await self._sio_client.connect(url, namespace=self._toplvl_namespace)
+
+    async def disconnect(self):
+        await self._sio_client.disconnect()
+
+    async def join(self, room_id: str):
+        await self._sio_client.emit(
+            "init", data={"client": self._sio_client.sid, "room": room_id}
+        )
+
+    async def leave(self, room_id: str):
+        await self._sio_client.emit(
+            "deinit", data={"client": self._sio_client.sid, "room": room_id}
+        )
+
+    async def chat(self, room_id: str, msg: str):
+        item = {"nickname": self._nickname, "msg": msg}
+        data = {"client": self._sio_client.sid, "gameID": room_id}
+        data.update(item)
+        await self._sio_client.emit("chat", data=data)
+        self._msg_log.append(item)
+
+    async def verify_join(self, expect_clients: List):
+        expect_client_sid = [v.sio.sid for v in expect_clients]
+        actual_client_sid = []
+        for expect_sid in expect_client_sid:
+            evts = await self._sio_client.receive(timeout=1)
+            assert len(evts) == 2
+            assert evts[0] == "init"
+            assert evts[1]["succeed"]
+            actual_client_sid.append(evts[1]["client"])
+        assert set(actual_client_sid) == set(expect_client_sid)
+
+    async def verify_chat(self, expect_sender, expect_error: Dict = None):
+        evts: List = await self._sio_client.receive(timeout=3)
+        assert len(evts) == 2
+        assert evts[0] == "chat"
+        if expect_error:
+            assert evts[1] == expect_error
+            self.messages_log.pop()
+        else:
+            assert evts[1]["client"] == expect_sender.sio.sid
+            assert evts[1]["nickname"] == expect_sender.nickname
+            item = {"nickname": evts[1].pop("nickname"), "msg": evts[1].pop("msg")}
+            self._msg_log.append(item)
+            assert self.messages_log == expect_sender.messages_log
+
+
+class TestRealTimeComm:
```

Comment:
可增加更多測試條件，例如重複加入`room`, 離開未加入的`room`,

Reply:
We didn't discuss this design issue in detail yet.
I'd prefer to deal with the design issue in my next PR (to avoid this PR from growing too large)

Here is the design approach
- `socket.io` server **needs to know** which game rooms are created, which players are currently in which room
- this is for detecting re-entrants , as well as the players who have never joined a game room
- feasible implementation , how can `socket.io` server knows about it ? : 
  - `socket.io` server maintains some kind of table map, and  the table map records events emitted from REST server on **creating game**
  - `socket.io` server directly read from [game repository](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/blob/main/backend/app/adapter/repository/__init__.py) through the abstract interface.  

Remind that :
  - GaaS Lobby platform creates a new game , by sending request to the REST endpoint `/games` , with `POST` method and list of player IDs
  - currently we (this game dev team) don't consider to let more players (or late comers) join an existing game room

Reply:
> I'd prefer to deal with the design issue in my next PR (to avoid this PR from growing too large)

贊成

>feasible implementation , how can socket.io server knows about it ? :
>- socket.io server maintains some kind of table map, and the table map records events emitted from REST server on creating game

我偏向這個選擇，理由如下：
1. [鬆耦合](https://zh.wikipedia.org/zh-tw/%E6%9D%BE%E8%80%A6%E5%90%88)性更好。通過事件通知的方式,REST API伺服器和Socket.IO伺服器彼此獨立,互不影響,更容易修改和擴展。
2. 分工清晰。REST API伺服器專注處理遊戲業務,Socket.IO伺服器專注處理即時通訊,各司其職,符合高內聚低耦合的設計原則。
3. 職責獨立。Socket.IO伺服器不需要直接訪問遊戲資料庫,只需要處理好事件通知即可,職責更加單一和獨立。
4. 更容易維護。業務邏輯獨立於即時通訊機制,即使遊戲業務規則變更,也不一定需要修改Socket.IO伺服器。